mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -277,17 +278,39 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
     val srcBlocks = blockify(rank, srcFeatures)
     val dstBlocks = blockify(rank, dstFeatures)
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
-      case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
-        val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
-        ratings.foreachActive { (i, j, r) =>
-          output(k) = (srcIds(i), (dstIds(j), r))
-          k += 1
-        }
-        output.toSeq
+      case (users, items) =>
+      val m = users.size
+      val n = math.min(items.size, num)
+      val output = new Array[(Int, (Int, Double))](m * n)
+      var j = 0
+      users.foreach (user => {
+          def order(a: (Int, Double)) = a._2
+          val pq: BoundedPriorityQueue[(Int, Double)] =
+            new BoundedPriorityQueue[(Int, Double)](n)(Ordering.by(order))
+          items.foreach (item => {
+              /**
+               * blas.ddot (F2jBLAS) is the same performance with the following code.
+               * the performace of blas.ddot with NativeBLAS is very bad.
+               * blas.ddot (F2jBLAS) is about 10% improvement comparing with linalg.dot.
+               * val rate = blas.ddot(rank, user._2, 1, item._2, 1)
+               */
+              var rate: Double = 0
+              var k = 0
+              while(k < rank) {
+                rate += user._2(k) * item._2(k)
```

Comment:
Then we can have `rate += srcFactor(k) * dstFactor(k)`

Also, can we call it `score` or `prediction` rather than `rate`?